src/ui/__tests__/ui_app.js

```diff
@@ -0,0 +1,263 @@
+import expect from 'expect.js';
+import UiApp from '../ui_app.js';
+import UiExports from '../ui_exports';
+import { isUndefined, noop } from 'lodash';
+
+function getMockSpec(extraParams) {
+  return {
+    id: 'uiapp-test',
+    main: 'main.js',
+    title: 'UIApp Test',
+    order: 9000,
+    description: 'Test of UI App Constructor',
+    icon: 'ui_app_test.svg',
+    linkToLastSubUrl: true,
+    hidden: false,
+    listed: null,
+    templateName: 'ui_app_test',
+    ...extraParams
+  };
+}
+describe('UiApp', () => {
+  describe('constructor', () => {
+    const uiExports = new UiExports({});
+
+    it('throw if ID is not given', () => {
+      function newAppMissingID() {
+        const spec = {}; // should have id property
+        const newApp = new UiApp(uiExports, spec);
+        return newApp;
+      }
+      expect(newAppMissingID).to.throwException();
+    });
+
+    describe('defaults', () => {
+      const spec = { id: 'uiapp-test-defaults' };
+      let newApp;
+      beforeEach(() => {
+        newApp = new UiApp(uiExports, spec);
+      });
+
+      it('use the spec ID', () => {
+        expect(newApp.id).to.be('uiapp-test-defaults');
+      });
+
+      it('creates fields with undefined value', () => {
+        // test that the fields exist, but have undefined value
+        expect('main' in newApp).to.be(true);
+        expect(isUndefined(newApp.main)).to.be(true);
+
+        expect('title' in newApp).to.be(true);
+        expect(isUndefined(newApp.title)).to.be(true);
+
+        expect('description' in newApp).to.be(true);
+        expect(isUndefined(newApp.description)).to.be(true);
+
+        expect('icon' in newApp).to.be(true);
+        expect(isUndefined(newApp.icon)).to.be(true);
+
+        expect('linkToLastSubUrl' in newApp).to.be(true);
+        expect(isUndefined(newApp.linkToLastSubUrl)).to.be(true);
```

Comment:
I deliberately didn't do that because it wouldn't actually assure anything. In JS, a field that was never set on an object evaluates to `undefined`

```
> const fooObj = {};
undefined
> fooObj.bar === undefined
true
```

The `in` operator was the only way I know of to test that a field has actually been set:
```
> 'bar' in fooObj
false
> fooObj.bar = 1;
1
> 'bar' in fooObj
true
```